src/types/package.rs

```diff
@@ -450,6 +482,107 @@ pub struct PackageDescriptor {
     pub package_type: PackageType,
 }
 
+/// Describes a package descriptor in the system with lockfile path
+#[derive(
+    PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug, Serialize, Deserialize, JsonSchema,
+)]
+pub struct LockFilePackageDescriptor {
+    pub package_descriptor: PackageDescriptor,
+    pub lockfile_path: Option<String>,
+}
+
+impl From<&PackageDescriptor> for LockFilePackageDescriptor {
+    fn from(value: &PackageDescriptor) -> Self {
+        LockFilePackageDescriptor {
+            package_descriptor: value.clone(),
+            lockfile_path: None,
+        }
+    }
+}
+
+/// Describes a package specifier in the system with lockfile path
+#[derive(
+    PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug, Serialize, Deserialize, JsonSchema,
+)]
+pub struct LockFilePackageSpecifier {
+    pub package_specifier: PackageSpecifier,
+    pub lockfile_path: Option<String>,
+}
+
+impl From<&PackageSpecifier> for LockFilePackageSpecifier {
+    fn from(value: &PackageSpecifier) -> Self {
+        LockFilePackageSpecifier {
+            package_specifier: value.clone(),
+            lockfile_path: None,
+        }
+    }
+}
+
+/// Describes a package url in the system with lockfile path
+#[derive(
+    PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug, Serialize, Deserialize, JsonSchema,
+)]
+pub struct ParsedPurl {
+    pub purl: String,
+    pub lockfile_path: Option<String>,
+}
+
+impl TryFrom<&PackageDescriptor> for ParsedPurl {
+    type Error = purl::PackageError;
+
+    fn try_from(value: &PackageDescriptor) -> Result<Self, Self::Error> {
+        let package_type = purl::PackageType::from(value.package_type);
+        let purl = Purl::builder(package_type, &value.name)
+            .with_version(&value.version)
+            .build()?
+            .to_string();
+        Ok(ParsedPurl {
+            purl,
+            lockfile_path: None,
+        })
+    }
+}
+
+impl TryFrom<&ParsedPurl> for PackageDescriptor {
+    type Error = purl::PackageError;
+
+    fn try_from(value: &ParsedPurl) -> Result<Self, Self::Error> {
+        let purl = Purl::from_str(&value.purl)?;
+        let package_type = PackageType::try_from(purl.package_type())?;
+
+        let package_name = match purl.namespace() {
+            Some(namespace) => {
+                let name = purl.name();
+                if package_type == PackageType::Maven {
+                    let ns = namespace.replace('/', ".");
+                    format!("{ns}:{name}")
+                } else {
+                    format!("{namespace}/{name}")
+                }
```

Comment:
Wouldn't it make sense for this to be handled in purl?

Reply:
Correct, at this point, I don't have a good way, without trying to parse though the package name and trying to figure out the namespace. I figured to let the purl crate handle that.